go-controller/pkg/controller/controller.go

```diff
@@ -214,3 +237,97 @@ func (c *controller[T]) ReconcileAll() {
 		c.queue.Add(key)
 	}
 }
+
+// SubController is a controller that should be used as a part of MultiController.
+// It is very similar to Controller, but it doesn't have its own initialSync method.
+// Initial sync is handled by MultiController instead.
+// SubController is not meant to be used standalone, all its methods are private and will be called by MultiController.
+type SubController interface {
+	addHandler() error
+	startWorkers(threadiness int) error
+	stop()
+}
+
+func NewSubController[T any](name string, config *Config[T]) SubController {
+	return &controller[T]{
+		name:   name,
+		config: config,
+		queue: workqueue.NewRateLimitingQueueWithConfig(
+			config.RateLimiter,
+			workqueue.RateLimitingQueueConfig{
+				Name: name,
+			},
+		),
+		stopChan: make(chan struct{}),
+		wg:       &sync.WaitGroup{},
+	}
+}
+
+// MultiController manages multiple controllers.
+// It is only needed when initial sync is required across multiple informers.
+// If your intialSync method `List`s more than one object type, it is very likely you should use MultiController.
+// InitialSync will be run after all subcontroller event handlers are added to ensure proper event handling
+// for related informers.
+type MultiController interface {
+	Start(threadiness int) error
+	Stop()
+}
+
+type multiController struct {
+	name        string
+	controllers []SubController
+	initialSync func() error
+}
+
+func NewMultiController(name string, initialSync func() error, controllers ...SubController) MultiController {
+	return &multiController{
+		name:        name,
+		controllers: controllers,
+		initialSync: initialSync,
+	}
+}
+
+// Start is a copy of Controller.Start, but it runs initialSync once and not per subcontroller.
+func (m *multiController) Start(threadiness int) (err error) {
+	klog.Infof("Starting multi-controller %v with %v workers", m.name, threadiness)
+
+	defer func() {
+		if err != nil {
+			m.Stop()
+		}
+	}()
+
+	g := new(errgroup.Group)
+	for _, subcontroller := range m.controllers {
+		subcontroller := subcontroller
+		g.Go(func() error {
+			return subcontroller.addHandler()
+		})
+	}
+	if err = g.Wait(); err != nil {
+		return err
+	}
+
+	// now we have already started receiving events and putting keys in the queue.
+	// If initial sync is needed, we can do it now, by listing all existing objects.
+	// Since we are receiving all events already, we know every change that happens after the following
+	// List call, will be handled.
+	if m.initialSync != nil {
+		if err = m.initialSync(); err != nil {
+			return fmt.Errorf("failed to start multi-controller %s: initial sync failed: %w", m.name, err)
+		}
+		klog.Infof("Multi-controller %s finished initial sync", m.name)
+	}
+	for _, subcontroller := range m.controllers {
+		if err = subcontroller.startWorkers(threadiness); err != nil {
```

Comment:
One of the main uses I would see is to share a thread pool for all controllers which is what one would think is actually happening. Instead we are spawning threadiness*len(subcontrollers) which is not all that apparent.